documentation/design/rar-core-scenarios.md

```diff
@@ -0,0 +1,193 @@
+The exact time the routine is executed depends on the latency of the asynchronous OS callback, on thread pool availability, CPU scheduling, and more.
+
+The focus of the following paragraphs is instead on SDK and NuGet references, because there are typically one to two orders of magnitude more of them than project
+references, so optimizing them has the best bang for the buck.
+
+### Obtain assembly names from the SDK
+
+The SDK is currently already passing relevant metadata such as `AssemblyVersion` and `PublicKeyToken`, so there is no need for RAR to open the file and parse its
+.NET metadata tables to get this information. This, together with the fact that SDK references are marked with `ExternallyResolved` so they cannot have dependencies
+outside of the primary set, means that there is no need to cache anything about these assemblies. Everything RAR needs comes (or can come if it's not there already)
+from the `Assemblies` parameter, explicitly provided on each invocation. Note, it may make sense to keep a cache in memory but it definitely doesn't make sense
+to save it to disk.
+
+If we do this, then in the warm and hot scenarios where the per project disk cache exists, we use it only to cache data about NuGet references and project references,
+significantly reducing its size. By eliminating per-reference I/O for most references, RAR would see a significant performance boost.
+
+This is assuming we trust the SDK that it passes correct data and we trust the user that they don't delete or overwrite their SDK files. If this assumption is not
+valid, the mitigation would be to store and check the timestamp of each individual file. We would still benefit from smaller on disk caches, being able to store only
+the timestamp and not assembly name for intact SDK references, but the hot scenario wouldn't get any faster than today.
+
+### Treat NuGet references as immutable
+
+NuGet references live in the NuGet cache which is conceptually immutable. If RAR takes advantage of this, it can eliminate timestamp checks for NuGet references as
+well. The risk is higher than for SDK references because overwriting files in the NuGet cache is commonly used as a cut-the-corner workaround. The benefit is smaller
+because the number of NuGet references is typically lower. The proposal is to use the immutability assumption by default, and provide an opt-out mechanism, likely in
+the form of a new `PackageReference` metadatum for the user to indicate that they plan to update references in the package.
+
+### Don't load the per project disk cache when not needed
+
+As described above, the on disk cache is not adding any value in the hot scenario because its contents already lives in the in-memory cache. The proposal is to
+load it lazily only when (and if) RAR runs into an assembly that does not have a record in the in-memory cache. In developer inner loop, when the same solution is
```

Comment:
Sorry for being confusing; I agree, but I was wondering if we could optimize that further but think that my proposed optimization would be a regression.

Only saving relevant assembly information is interesting. I could see it helping or hurting. Here's what I'm thinking:
Good:
Smaller footprint on disk
If we decide we need to load the cache (because we saw a cache miss), that load is faster and more targeted
If we decide we need to serialize to disk, that serialization is faster because there are fewer things in it

Bad:
If we decide we need to serialize to disk, that serialization is slower because we need to do some processing to figure out what we're actually serializing
If the node is used for multiple projects again (as is likely), there's a higher chance of a cache miss leading to deserializing another cache. So like if we have projects A and B with disjoint assembly dependencies, and we put one of B's dependencies in A's cache, then when we call RAR on A, we load B's dependency, and when we call RAR on B, we don't have to go back to disk to retrieve that dependency from B's cache. If this is common, it would outweigh the benefit of faster cache deserialization, since computers do well with bulk operations, not sequences of quick disk accesses.
Makes RAR's caches a bit more complicated (minor)

Overall, I'm not sure how to weigh those against each other. The footprint on disk seems like the biggest win to me, and I _think_ that would outweigh the other factors in my view, but I could easily be wrong.

Reply:
> Overall, I'm not sure how to weigh those against each other. The footprint on disk seems like the biggest win to me, and I think that would outweigh the other factors in my view, but I could easily be wrong.

+1
And the more I think about it, the more weight I put on deterministic behavior.

The downside with potentially having to read every project's cache (unless one is a superset of another and happens to be loaded first) only applies to cold builds, too.